roadmap/implementers-guide/src/node/disputes/dispute-distribution.md

```diff
@@ -1,3 +1,261 @@
+received a `SendDispute` message for that candidate.
+
+## Backing and Approval Votes
+
+Backing and approval votes get imported when they arrive/are created via the
+distpute coordinator by corresponding subsystems.
+
+We assume that under normal operation each node will be aware of backing and
+approval votes and optimize for that case. Nevertheless we want disputes to
+conclude fast and reliable, therefore if a node is not aware of backing/approval
+votes it can request the missing votes from the node that informed it about the
+dispute.
+
+## Resiliency
+
+The above protocol should be sufficient for most cases, but there are certain
+cases we also want to have covered:
+
+- Non validator nodes might be interested in ongoing voting, even before it is
+  recorded on chain.
+- Nodes might have missed votes, especially backing or approval votes.
+  Recovering them from chain is difficult and expensive, due to runtime upgrades
+  and untyped extrinsics.
+
+To cover those cases, we introduce a second request/response protocol, which can
+be handled on a lower priority basis as the one above. It consists of the
+request/response messages as described in the [protocol
+section][#vote-recovery].
+
+Nodes may send those requests to validators, if they feel they are missing
+votes. E.g. after some timeout, if no majority was reached yet in their point of
+view or if they are not aware of any backing/approval votes for a received
+disputed candidate.
+
+The receiver of a `IHaveVotesRequests` message will do the following:
+
+1. See if the sender is missing votes we are aware of - if so, respond with
+   those votes. Also send votes of equivocating validators, no matter the
+   bitfield.
+2. Check whether the sender knows about any votes, we don't know about and if so
+   send a `IHaveVotes` request back, with our knowledge.
+3. Record the peer's knowledge.
+
+When to send `IHaveVotes` messages:
+
+1. Whenever we are asked to do so via
+   `DisputeDistributionMessage::FetchMissingVotes`.
+2. Approximately once per block to some random validator as long as the dispute
+   is active.
+
+Spam considerations: Nodes want to accept those messages once per validator and
+per slot. They are free to drop more frequent requests or requests for stale
+data. Requests coming from non validator nodes, can be handled on a best effort
+basis.
+
+## Considerations
+
+Dispute distribution is critical. We should keep track of available validator
+connections and issue warnings if we are not connected to a majority of
+validators. We should also keep track of failed sending attempts and log
+warnings accordingly. As disputes are rare and TCP is a reliable protocol,
+probably each failed attempt should trigger a warning in logs and also logged
+into some Prometheus metric.
+
+## Disputes for non included candidates
+
+If deemed necessary we can later on also support disputes for non included
+candidates, but disputes for those cases have totally different requirements.
```

Comment:
It's not described anywhere in the document how a receiving node can tell whether a dispute is on an included or non-included candidate

Reply:
The only way I am aware of is to iterate all heads of the chain, traversing backwards to the genesis, and look for `CandidateIncluded` events.

This has problems we've discussed at length before:
1. State pruning. Events for old blocks aren't necessarily available
2. Expense. It's an expensive computation
3. Reversions. When we revert some bad candidate we might begin to think that the candidate wasn't included anymore, when it's crucial that the dispute be transplanted onto other branches

Reply:
Sorry, I still mean by included, that it has been included somewhere - thus it is available. Changed it to say "available" instead.